src/data-workspace/use-data-value-mutation.js

```diff
@@ -0,0 +1,95 @@
+    const attributeOptionComboId = useAttributeOptionCombo()
+    const engine = useDataEngine()
+
+    const dataValueQueryKey = [
+        dataValueQuery,
+        {
+            dataSetId,
+            periodId,
+            orgUnitId,
+            attributeOptionComboId,
+        },
+    ]
+    const mutationFn = (variables) =>
+        engine.mutate(DATA_VALUE_MUTATION, { variables })
+
+    return useMutation(mutationFn, {
+        // Optimistic update of the react-query cache
+        onMutate: async (newDataValue) => {
+            // Cancel any outgoing refetches (so they don't overwrite our optimistic update)
+            await queryClient.cancelQueries(dataValueQueryKey)
+
+            // Snapshot the previous value
+            const previousDataValues =
+                queryClient.getQueryData(dataValueQueryKey)
+
+            // Optimistically update to the new value
+            queryClient.setQueryData(dataValueQueryKey, () => {
+                const newDataValues =
+                    previousDataValues.dataValues.dataValues.map(
+                        (dataValue) => {
+                            const {
+                                categoryOptionCombo,
+                                dataElement,
+                                orgUnit,
+                                period,
+                            } = dataValue
+                            const { co, de, ou, pe, value } = newDataValue
+                            const match =
+                                categoryOptionCombo === co &&
+                                dataElement === de &&
+                                orgUnit === ou &&
+                                period === pe
+
+                            if (!match) {
+                                return dataValue
+                            }
+
+                            return {
+                                ...dataValue,
+                                value,
+                            }
+                        }
+                    )
+
+                // Ensure we don't mutate previousDataValues
+                return {
+                    ...previousDataValues,
+                    dataValues: {
+                        ...previousDataValues.dataValues,
+                        dataValues: newDataValues,
+                    },
+                }
+            })
+
+            return { previousDataValues, dataValueQueryKey }
+        },
+        // If the mutation fails, use the context returned from onMutate to roll back
+        onError: (err, newDataValue, context) => {
+            queryClient.setQueryData(
+                context.dataValueQueryKey,
+                context.previousDataValues
+            )
+        },
+        // Always refetch after error or success
+        // eslint-disable-next-line max-params
+        onSettled: (newDataValue, error, variables, context) => {
+            queryClient.invalidateQueries(context.dataValueQueryKey)
```

Comment:
So with regards to the refetch & rollback after an error, that's to address this:

> In most of these failure cases, you can just trigger a refetch for your optimistic queries to revert them to their true server state. In some circumstances though, refetching may not work correctly and the mutation error could represent some type of server issue that won't make it possible to refetch. In this event, you can instead choose to rollback your update.

Reply:
In a similar vein, just keeping the optimistic update on success is a possibility. Syncing with the server after success is just a way to ensure that our data is 100% consistent with the backend, as for example our optimistic updates for new dataValues aren't completely realistic (they lack certain bits of data that we can't predict, like creation time, etc.).

Reply:
Oh and to answer:

> And I guess this doesn't happen when you are offline, only when the server rejects or someting?

That's correct, mutations will only succeed or fail once the user is online.

Reply:
I see. I actually think that we don't want to rollback, as we should also integrate this with validations. So to prevent confusing scenarios (eg. multiple mutations in a row, that rollbacks - like you said), I think that showing errors in a good way and let the user fix the error is a better solution.
But yeah, we could try it out and remove it later.